corehq/ex-submodules/soil/progress.py

```diff
@@ -77,6 +79,50 @@ def set_task_progress(task, current, total):
     update_task_state(task, 'PROGRESS', {'current': current, 'total': total})
 
 
+class ProgressManager(object):
+    """
+    A context manager that mediates calls to `set_task_progress`
+
+    and inserts exponential backoff (with a max interval) to spare the db
+    and flushes on __exit__
+
+    """
+    def __init__(self, task, min_interval=.5, exp_rate=1.20, max_interval=10 * 60):
+        self.task = task
+        self._interval = min_interval
+        self._exp_rate = exp_rate
+        self._max_interval = max_interval
+        self._next_flush_time = datetime.datetime.utcnow()
+        self._value = {'current': None, 'total': None}
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self.flush()
+
+    def set_progress(self, current, total):
+        new_value = {'current': current, 'total': total}
+
+        if new_value != self._value:
+            self._value = new_value
+            if self._should_flush():
+                self.flush()
+                self._set_next_flush_time()
+
+    def _should_flush(self):
+        return self._next_flush_time <= datetime.datetime.utcnow()
+
+    def flush(self):
+        self._set_task_progress(self.task, **self._value)
+
+    def _set_next_flush_time(self):
+        self._next_flush_time += datetime.timedelta(seconds=self._interval)
```

Comment:
Should this be `self._next_flush_time = datetime.utcnow() +delta`? That would guarantee that next flush time is always in the future.

Reply:
Maybe I should leave a comment here. We have a choice like you say of waiting some interval since _the last flush_ (as you suggested) versus having a predetermined schedule of flushes, or more precisely the max number of flushes allowed to have happened before a given time X, which is what I chose to do here.

The benefit of the predetermined schedule is that it doesn't punish processes that have a super slow rate of calls to `progress_manager.set_progress`, i.e. pinging that like crazy versus pinging it once a minute will let the same number of calls through to the db in a given amount of time (provided both call patterns eventually reach the max of the time period). For a relatively short total time, this may mean not throttling the one with a slow rate of calls to `progress_manager.set_progress` at all. This strategy overall seemed most likely to throttle where throttling would save us a lot, and not further reduce the resolution on processes that already have low resolution progress bars (with little overall benefit to db calls).